exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/managed/ExternalSortBatch.java

```diff
@@ -0,0 +1,1321 @@
+    // batches (the minimum needed to make progress) and the allowance for the
+    // output batch.
+
+    long minLoadMemory = spillPoint + estimatedInputSize;
+
+    // The merge memory pool assumes we can spill all input batches. To make
+    // progress, we must have at least two merge batches (same size as an output
+    // batch) and one output batch. Again, double to allow for power-of-two
+    // allocation and add one for a margin of error.
+
+    long minMergeMemory = (2*3 + 1) * estimatedOutputBatchSize;
+
+    // Determine how much memory can be used to hold in-memory batches of spilled
+    // runs when reading from disk.
+
+    mergeMemoryPool = Math.max(minMergeMemory,
+                               (long) ((memoryLimit - 3 * estimatedOutputBatchSize) * 0.95));
+
+    // Sanity check: if we've been given too little memory to make progress,
+    // issue a warning but proceed anyway. Should only occur if something is
+    // configured terribly wrong.
+
+    long minMemoryNeeds = Math.max(minLoadMemory, minMergeMemory);
+    if (minMemoryNeeds > memoryLimit) {
+      logger.warn("updateMemoryEstimates: potential memory overflow! " +
+                   "Minumum needed = {} bytes, actual available = {} bytes",
+                   minMemoryNeeds, memoryLimit);
+    }
+
+    // Log the calculated values. Turn this on if things seem amiss.
+    // Message will appear only when the values change.
+
+    logger.debug("Memory Estimates: record size = {} bytes; input batch = {} bytes, {} records; " +
+                  "output batch size = {} bytes, {} records; " +
+                  "Available memory: {}, spill point = {}, min. merge memory = {}",
+                estimatedRecordSize, estimatedInputBatchSize, estimatedInputRecordCount,
+                estimatedOutputBatchSize, outputBatchRecordCount,
+                memoryLimit, spillPoint, minMergeMemory);
+  }
+
+  /**
+   * Determine if spill is needed before receiving the new record batch.
+   * Spilling is driven purely by memory availability (and an optional
+   * batch limit for testing.)
+   *
+   * @return true if spilling is needed, false otherwise
+   */
+
+  private boolean isSpillNeeded() {
+
+    // Can't spill if less than two batches else the merge
+    // can't make progress.
+
+    if (bufferedBatches.size() < 2) {
+      return false; }
+
+    // Must spill if we are below the spill point (the amount of memory
+    // needed to do the minimal spill.)
+
+    long freeMemory = memoryLimit - allocator.getAllocatedMemory();
+
+    // Sanity check. If the memory goes negative, the calcs are off.
+
+    if (freeMemory < 0) {
+      logger.error("ERROR: Free memory is negative: {}. Spill point = {}",
+                   freeMemory, spillPoint);
+    }
+    if (freeMemory <= spillPoint) {
+      return true; }
+
+    // Number of incoming batches (BatchGroups) exceed the limit and number of incoming
+    // batches accumulated since the last spill exceed the defined limit
```

Comment:
Stale comment about my attempt to understand the original implementation. Replaced with a better comment about the actual current implementation.